src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,545 @@
+                    ExpressionSyntax parent when !originalReceiver.IsKind(SyntaxKind.ConditionalAccessExpression) => parent,
+                    // Otherwise, we rewrite the original receiver with the expression on the left-hand-side of this name.
+                    // For instance, if we had `a?.b.c && a?.b.d` we rewrite it as `a?.b is { c: true, d: false }`
+                    MemberAccessExpressionSyntax memberAccess => originalReceiver.ReplaceNode(memberAccess, memberAccess.Expression),
+                    // Similarly, if we had `a?.b && a?.c` we rewrite it as `a is { b: true, c: true }`
+                    MemberBindingExpressionSyntax { Parent: ConditionalAccessExpressionSyntax p } => p.Expression,
+                    // There should be no other possibility as we have walked downwards already and verified each node in the path.
+                    var v => throw ExceptionUtilities.UnexpectedValue(v),
+                };
+            }
+
+            static ExpressionSyntax? SkipCommonNames(ArrayBuilder<IdentifierNameSyntax> leftNames, ArrayBuilder<IdentifierNameSyntax> rightNames)
+            {
+                ExpressionSyntax? lastName = null;
+                int leftIndex, rightIndex;
+                for (leftIndex = leftNames.Count - 1, rightIndex = rightNames.Count - 1; leftIndex > 0 && rightIndex > 0; leftIndex--, rightIndex--)
+                {
+                    var leftName = leftNames[leftIndex];
+                    var rightName = rightNames[rightIndex];
+                    if (!AreEquivalent(leftName, rightName))
+                        break;
+                    lastName = leftName;
+                }
+
+                leftNames.Clip(leftIndex + 1);
+                rightNames.Clip(rightIndex + 1);
+                return lastName;
+            }
+        }
+
+        private static ExpressionSyntax? GetInnermostReceiver(ExpressionSyntax node, out ImmutableArray<IdentifierNameSyntax> namesOpt, SemanticModel model)
+        {
+            using var _ = ArrayBuilder<IdentifierNameSyntax>.GetInstance(out var builder);
+            var receiver = GetInnermostReceiver(node, builder, model);
+            namesOpt = builder.ToImmutableOrNull();
+            return receiver;
+        }
+
+        private static ExpressionSyntax? GetInnermostReceiver(ExpressionSyntax node, ArrayBuilder<IdentifierNameSyntax> builder, SemanticModel model)
+        {
+            return GetInnermostReceiver(node);
+
+            ExpressionSyntax? GetInnermostReceiver(ExpressionSyntax node)
+            {
+                switch (node)
+                {
+
+                    case IdentifierNameSyntax name
+                    when CanConvertToSubpattern(name):
+                        builder.Add(name);
+                        // This is a member reference with an implicit `this` receiver.
+                        return null;
```

Comment:
i don't love this assumption.  it feels a bit non-rigorous.

Reply:
We know this is true because we already checked CanConvertToSubpattern. I'll add a comment and assert IOp.